scipy/sparse/linalg/isolve/iterative.py

```diff
@@ -220,214 +412,174 @@ def bicg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None
+    if atol is not None:
+        tol = max(float(atol)/r0norm, tol)
+    for iter in range(maxiter):
+        rho1 = np.inner(rhat.conjugate(), r)
+        if rho1 == 0.:
+            print("Iterative method failed due to inner product of residuals is zero")
+        if iter == 0:
+            p = r.copy()
+        else:
+            beta = (rho1/rho0) * (alpha/omega)
+            p = r + beta * (p - omega*v)
+        p_hat = M.matvec(p)
+        v = A.matvec(p_hat)
+        rhattv = np.inner(rhat.conjugate(), v)
+        if rhattv == 0.:
+            print("Warning: divide by zero (in step {})".format(iter+1))
+            return (postprocess(x), -1)
+        alpha = rho1 / rhattv
+        s = r - alpha * v
+        if np.linalg.norm(s) < 1e-14:
+            x += alpha * p_hat
+            return (postprocess(x), 0)
+        s_hat = M.matvec(s)
+        t = A.matvec(s_hat)
+        omega = np.inner(t.conjugate(), s) / np.inner(t.conjugate(), t)
+        x += alpha * p_hat + omega * s_hat
+        r = s - omega * t
+        beta = np.linalg.norm(r) / r0norm
+        if callback is not None:
             callback(x)
-        slice1 = slice(ndx1-1, ndx1-1+n)
-        slice2 = slice(ndx2-1, ndx2-1+n)
-        if (ijob == -1):
-            if callback is not None:
-                callback(x)
-            break
-        elif (ijob == 1):
-            work[slice2] *= sclr2
-            work[slice2] += sclr1*matvec(work[slice1])
-        elif (ijob == 2):
-            work[slice1] = psolve(work[slice2])
-        elif (ijob == 3):
-            work[slice2] *= sclr2
-            work[slice2] += sclr1*matvec(x)
-        elif (ijob == 4):
-            if ftflag:
-                info = -1
-                ftflag = False
-            resid, info = _stoptest(work[slice1], atol)
-        ijob = 2
-
-    if info > 0 and iter_ == maxiter and not (resid <= atol):
-        # info isn't set appropriately otherwise
-        info = iter_
-
-    return postprocess(x), info
+        if beta < tol:
+            return (postprocess(x), 0)
+        elif iter == maxiter - 1:
+            return (postprocess(x), maxiter)
+        rho0 = rho1;
 
 
-@set_docstring('Use Conjugate Gradient iteration to solve ``Ax = b``.',
+#======================
+#   CGNR
+#======================
+@set_docstring('Use Conjugate Gradient Normal Residual iteration to solve '
+               '``Ax = b``.',
                'The real or complex N-by-N matrix of the linear system.\n'
-               '``A`` must represent a hermitian, positive definite matrix.\n'
                'Alternatively, ``A`` can be a linear operator which can\n'
                'produce ``Ax`` using, e.g.,\n'
                '``scipy.sparse.linalg.LinearOperator``.')
 @non_reentrant()
-def cg(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None, atol=None):
-    A, M, x, b, postprocess = make_system(A, M, x0, b)
-
-    n = len(b)
+def cgnr(A, b, x0=None, tol=1e-5, maxiter=None, M=None, callback=None):
```

Comment:
I think changing the name from `cg` to `cgnr` is backward incompatible and it also looks like `atol` is no longer in the function's signature which will likely break existing code for users.

Reply:
Yes, the rule is simple: don't change any existing names, the only thing you can do is add new keywords at the end of a function signature.

Reply:
Dear @zoj613 @rgommers 
Hm, I am sorry for reordering the function names according to the history of these methods rather than original alphabet. I'll reorder the function names of iterative methods according to alphabet in the `iterative.py`.
CG is reimplemented (As your suggestions earlier, the option `atol` has not been removed to keep compatibility). The new keyword `cgnr` is an implimentation for Conjugate Gradient Normal Residual, which is a variant of CG (it is used to solve asymmetric problems). I added references cited and will re-submit this code to get rid to the confusion. Thanks